vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt

```diff
@@ -0,0 +1,119 @@
+    @Inject lateinit var session: Session
+
+    override fun getLayoutResId() = R.layout.fragment_qr_code_scanner
+
+    private val createDirectRoomViewModel: CreateDirectRoomViewModel by activityViewModel()
+
+    @CallSuper
+    override fun injectWith(injector: ScreenComponent) {
+        session = injector.activeSessionHolder().getActiveSession()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        // Register ourselves as a handler for scan results.
+        scannerView.setResultHandler(this)
+        // Start camera on resume
+        scannerView.startCamera()
+    }
+
+    override fun onPause() {
+        super.onPause()
+        // Stop camera on pause
+        scannerView.stopCamera()
+    }
+
+    // Copied from https://github.com/markusfisch/BinaryEye/blob/
+    // 9d57889b810dcaa1a91d7278fc45c262afba1284/app/src/main/kotlin/de/markusfisch/android/binaryeye/activity/CameraActivity.kt#L434
+    private fun getRawBytes(result: Result): ByteArray? {
+        val metadata = result.resultMetadata ?: return null
+        val segments = metadata[ResultMetadataType.BYTE_SEGMENTS] ?: return null
+        var bytes = ByteArray(0)
+        @Suppress("UNCHECKED_CAST")
+        for (seg in segments as Iterable<ByteArray>) {
+            bytes += seg
+        }
+        // byte segments can never be shorter than the text.
+        // Zxing cuts off content prefixes like "WIFI:"
+        return if (bytes.size >= result.text.length) bytes else null
+    }
+
+    private fun addByQrCode(value: String) {
+        val mxid = (PermalinkParser.parse(value) as? PermalinkData.UserLink)?.userId
+
+        if (mxid === null) {
+            Toast.makeText(requireContext(), R.string.invalid_qr_code_uri, Toast.LENGTH_SHORT).show()
+            requireActivity().finish()
+        } else {
+            // This and the related conditional can be removed when PR #2342 is merged
+            val existingDm = session.getExistingDirectRoomWithUser(mxid)
+
+            if (existingDm === null) {
+                // The following assumes MXIDs are case insensitive
+                if (mxid.equals(other = session.myUserId, ignoreCase = true)) {
+                    Toast.makeText(requireContext(), R.string.cannot_dm_self, Toast.LENGTH_SHORT).show()
+                    requireActivity().finish()
+                } else {
+                    // Try to get user from known users and fall back to creating a User object from MXID
+                    val qrInvitee = if (session.getUser(mxid) != null) session.getUser(mxid)!! else User(mxid, null, null)
+
+                    createDirectRoomViewModel.handle(
+                            CreateDirectRoomAction.CreateRoomAndInviteSelectedUsers(setOf(PendingInvitee.UserPendingInvitee(qrInvitee)))
+                    )
+                }
+            } else {
+                navigator.openRoom(requireContext(), existingDm, null, false)
+                requireActivity().finish()
+            }
+        }
```

Comment:
You should move this code (all the `else` block) in the viewModel, by adding an action in `CreateDirectRoomAction`

Reply:
I think the whole if/else can be removed if/when #2342 gets merged, right? In that case creating a new DM will automatically switch to the old one anyway, so there will be no need to check that here. Should I still move the else block or wait for #2342 to get merged, rebase, and then remove the whole thing?

Reply:
#2342 will be merged soon, I will see after what we can do here.

Reply:
I've just updated my branch and removed the whole if statement. It seems to work as desired.